src/gf.c

```diff
@@ -1504,6 +1588,20 @@ JL_DLLEXPORT jl_value_t *jl_apply_generic(jl_value_t **args, uint32_t nargs)
             // unreachable
         }
     }
+    else {
+        mfunc = entry->func.linfo;
+    }
+    jl_method_t *m = mfunc->def;
+    if (m->ambig != jl_nothing) {
+        tt = arg_type_tuple(args, nargs);
+        jl_value_t *ambiguities = NULL;
+        JL_GC_PUSH2(&tt, &ambiguities);
+        ambiguities = jl_list_call_ambiguities_(tt, m);
+        if (ambiguities != jl_nothing) {
+            JL_GC_POP();
+            jl_ambiguous_method_error((jl_value_t*)tt, ambiguities);
```

Comment:
I think `AmbiguousMethodError` should have the same fields as `NoMethodError`. Constructing the full list of ambiguities seems like a job for error printing. At an exception throwing site I prefer to quickly package up the minimum information you need. If the error is going to be printed, there's plenty of time to dig through method tables.